ts/modules/broken-axis.src.ts

```diff
@@ -241,457 +158,626 @@ Axis.prototype.setBreaks = function (
-                if (newMax < newMin) {
-                    newMax = newMin;
+            while (i--) {
+                if (BrokenAxisAdditions.isInBreak((breaks as any)[i], val)) {
+                    inbrk = true;
+                    if (!keep) {
+                        keep = pick(
+                            (breaks as any)[i].showPoints,
+                            !axis.isXAxis
+                        );
+                    }
                 }
             }
-            Axis.prototype.setExtremes.call(
-                this,
-                newMin,
-                newMax,
-                redraw,
-                animation,
-                eventArguments
-            );
-        };
 
-        axis.setAxisTranslation = function (saveOld?: boolean): void {
-            Axis.prototype.setAxisTranslation.call(this, saveOld);
+            if (inbrk && testKeep) {
+                ret = inbrk && !keep;
+            } else {
+                ret = inbrk;
+            }
+        }
+        return ret;
+    }
 
-            this.unitLength = null as any;
-            if (this.isBroken) {
-                var breaks = axis.options.breaks,
-                    // Temporary one:
-                    breakArrayT = [] as Array<Highcharts.AxisBreakBorderObject>,
-                    breakArray = [] as Array<Highcharts.AxisBreakObject>,
-                    length = 0,
-                    inBrk: number,
-                    repeat: number,
-                    min = axis.userMin || axis.min,
-                    max = axis.userMax || axis.max,
-                    pointRangePadding = pick(axis.pointRangePadding, 0),
-                    start: (number|null|undefined),
-                    i: number;
-
-                // Min & max check (#4247)
-                (breaks as any).forEach(function (
-                    brk: Highcharts.XAxisBreaksOptions
-                ): void {
-                    repeat = brk.repeat || Infinity;
-                    if (axis.isInBreak(brk, min)) {
-                        (min as any) +=
-                            ((brk.to as any) % repeat) -
-                            ((min as any) % repeat);
+    /**
+     * Dynamically set or unset breaks in an axis. This function in lighter than
+     * usin Axis.update, and it also preserves animation.
+     *
+     * @private
+     * @function Highcharts.Axis#setBreaks
+     *
+     * @param {Array<Highcharts.XAxisBreaksOptions>} [breaks]
+     *        The breaks to add. When `undefined` it removes existing breaks.
+     *
+     * @param {boolean} [redraw=true]
+     *        Whether to redraw the chart immediately.
+     *
+     * @return {void}
+     */
+    public setBreaks(
+        breaks?: Array<Highcharts.XAxisBreaksOptions>,
+        redraw?: boolean
+    ): void {
+        const brokenAxis = this;
+        const axis = brokenAxis.axis;
+        const isBroken = (isArray(breaks) && !!breaks.length);
```

Comment:
As an equivalent to what we have now done with datetime axis and logarithmic axis, shouldn't `isBroken` be either removed so we solely look for `Axis.brokenAxis`, or alternatively be moved to `Axis.brokenAxis`?

Reply:
I am not sure about this. This is part of the broken axis composition to actually check, what to do, if there are no breaks. I will rename it to `hasBreaks`.